include/pybind11/cast.h

```diff
@@ -125,6 +243,158 @@ PYBIND11_NOINLINE inline handle get_type_handle(const std::type_info &tp, bool t
+        type_info *type() { return *typeit; }
+    };
+
+    iterator begin() { return iterator(inst); }
+    iterator end() { return iterator(inst, true); }
+
+    iterator find(const type_info *find_type) {
+        auto it = begin(), endit = end();
+        while (it != endit && it.type() != find_type) ++it;
+        return it;
+    }
+
+    value_and_holder front() { return *begin(); }
+
+    size_t size() { return inst->all_type_info().size(); }
+};
+
+/** Extracts C++ value and holder pointer references from an instance (which may contain multiple
+ * values/holders for python-side multiple inheritance) that match the given type.  Throws an error
+ * if the given type (or ValueType, if omitted) is not a pybind11 base of the given instance.
+ *
+ * The returned object should be short-lived: in particular, it must not outlive the called-upon
+ * instance.
+ */
+PYBIND11_NOINLINE inline value_and_holder instance::get_value_and_holder(const type_info *find_type) {
+    detail::values_and_holders vhs(this);
+    auto it = vhs.find(find_type);
+    if (it != vhs.end())
+        return *it;
+
+#if defined(NDEBUG)
+    pybind11_fail("pybind11::detail::instance::get_value_and_holder: "
+            "type is not a pybind11 base of the given instance "
+            "(compile in debug mode for type details)");
+#else
+    pybind11_fail("pybind11::detail::instance::get_value_and_holder: `" +
+            std::string(find_type->type->tp_name) + "' is not a pybind11 base of the given `" +
+            std::string(Py_TYPE(this)->tp_name) + "' instance");
+#endif
+}
+
+PYBIND11_NOINLINE inline void instance::allocate_layout() {
+    new (&all_type_info_storage) detail::all_type_info(this);
+
+    const size_t n_types = all_type_info().size();
+    simple_layout =
+        n_types == 1 && all_type_info().front()->holder_size_in_ptrs <= instance_simple_holder_in_ptrs();
+
+    // Simple path: no python-side multiple inheritance, and a small-enough holder
+    if (simple_layout) {
+        simple_value_holder[0] = nullptr;
+        simple_holder_constructed = false;
+    }
+    else { // multiple base types or a too-large holder
+        // Allocate space to hold: [bb...][v1*][h1][v2*][h2]... where [vN*] is a value pointer, [hN]
+        // is the (uninitialized) holder instance for value N, and [bb...] is a bitfield that tracks
+        // whether each associated holder has been initialized.  Each [block] is an integer multiple
+        // of sizeof(void *) (padded if necessary).
+        const size_t n_flag_ptrs = n_types / value_and_holder::bits_per_ptr +
+                                  (n_types % value_and_holder::bits_per_ptr != 0);
+        size_t space = n_flag_ptrs;
+        for (auto tinfo : all_type_info()) {
+            space += 1;
+            space += tinfo->holder_size_in_ptrs;
+        }
+
+        // Allocate space for flags, values, and holders:
+        values_and_holders = (void **) PyMem_New(void *, space);
```

Comment:
Python's `PyMem_New` wraps malloc itself, but also keeps its own pool of free memory, so it has the potential to be closer in memory to the PyObject itself.  It may perform better in a tight allocate-deallocate loop (with Python reusing memory rather than needing to dealloc/malloc).  It doesn't mix with C++ allocation via new/delete, but in this case, where we just need a small chunk of memory that is controlled entirely within pybind, it seems appropriate.